src/Interpreters/MonotonousOrderByVisitor.h

```diff
@@ -0,0 +1,194 @@
+#pragma once
+
+#include <DataTypes/DataTypeFactory.h>
+#include <Functions/FunctionFactory.h>
+#include <IO/WriteHelpers.h>
+#include <Interpreters/InDepthNodeVisitor.h>
+#include <Interpreters/IdentifierSemantic.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTOrderByElement.h>
+#include <Parsers/ASTSelectQuery.h>
+#include <Parsers/ASTTablesInSelectQuery.h>
+#include <Parsers/IAST.h>
+#include <Common/typeid_cast.h>
+
+namespace DB
+{
+
+class MonotonicityCheckMatcher
+{
+public:
+    struct Data
+    {
+        const TablesWithColumns & tables;
+        const Context & context;
+        ASTPtr & identifier;
+        bool & is_monotonous;
+        bool & is_positive;
+        DataTypePtr data_type = nullptr;
+        bool first = true;
+        bool done = false;
+    };
+
+    static void visit(const ASTPtr & ast, Data & data)
+    {
+        auto * ast_function = ast->as<ASTFunction>();
+        if (!data.done && ast_function)
+        {
+            visit(*ast_function, data);
+        }
+    }
+
+
+    static void visit(ASTFunction & ast_function, Data & data)
+    {
+        if (!data.is_monotonous)
+            return;
+
+        auto arguments = ast_function.arguments;
+        if (arguments->children.size() != 1)
+        {
+            data.is_monotonous = false;
+            data.done = true;
+            return;
+        }
+
+        if (data.first)
+            data.identifier = ast_function.arguments->children[0];
+
+        const auto & function = FunctionFactory::instance().tryGet(ast_function.name, data.context);
+        auto pos = IdentifierSemantic::chooseTable(*data.identifier->as<ASTIdentifier>(), data.tables, true);
+        auto data_type_and_name = data.tables[*pos].columns.tryGetByName(data.identifier->getAliasOrColumnName());
```

Comment:
`getAliasOrColumnName()` could return alias. It also could be qualified with table name: `t.x` instead of `x`. But you need not qualified column name here. Use `ASTIdentifier::shortName()` instead.